File: Sources/AlexaDeviceManager/Models/ApplianceID.swift
import Foundation

struct DecodedApplianceID: Hashable {
    let skillId: String
    let stage: String?
    let domain: String
    let objectId: String
}

enum ApplianceIDParser {
    private static let regex: NSRegularExpression = {
        try! NSRegularExpression(pattern: "^SKILL_([A-Za-z0-9+/=]+)_(.+)$")
    }()

    /// Decodes applianceIds in the format `SKILL_<base64-json>_<domain>#<object_id>`.
    /// Other skill types / native Amazon devices don't match this format -> nil.
    static func decode(_ applianceId: String) -> DecodedApplianceID? {
        let fullRange = NSRange(applianceId.startIndex..., in: applianceId)
        guard let match = regex.firstMatch(in: applianceId, range: fullRange),
              let base64Range = Range(match.range(at: 1), in: applianceId),
              let suffixRange = Range(match.range(at: 2), in: applianceId) else {
            return nil
        }

        let base64Part = String(applianceId[base64Range])
        let suffix = String(applianceId[suffixRange])

        guard let jsonData = Data(base64Encoded: base64Part),
              let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: String],
              let skillId = json["skillId"] else {
            return nil
        }

        let domainSplit = suffix.split(separator: "#", maxSplits: 1)
        let domain = domainSplit.first.map(String.init) ?? suffix
        let objectId = domainSplit.count > 1 ? String(domainSplit[1]) : ""

        return DecodedApplianceID(skillId: skillId, stage: json["stage"], domain: domain, objectId: objectId)
    }
}

File: app/models/appliance_id.py
from __future__ import annotations
import re
import base64
import json
from dataclasses import dataclass


@dataclass
class DecodedApplianceID:
    skill_id: str
    stage: str | None
    domain: str
    object_id: str


class ApplianceIDParser:
    _regex = re.compile(r"^SKILL_([A-Za-z0-9+/=]+)_(.+)$")

    @staticmethod
    def decode(appliance_id: str) -> DecodedApplianceID | None:
        match = ApplianceIDParser._regex.match(appliance_id)
        if not match:
            return None

        base64_part = match.group(1)
        suffix = match.group(2)

        try:
            json_data = base64.b64decode(base64_part)
            payload: dict[str, str] = json.loads(json_data)
            skill_id = payload.get("skillId")
            if not skill_id:
                return None
        except (json.JSONDecodeError, base64.binascii.Error, UnicodeDecodeError):
            return None

        domain_split = suffix.split("#", maxsplit=1)
        domain = domain_split[0] if domain_split else suffix
        object_id = domain_split[1] if len(domain_split) > 1 else ""

        return DecodedApplianceID(
            skill_id=skill_id,
            stage=payload.get("stage"),
            domain=domain,
            object_id=object_id,
        )
